### PR TITLE
Try: Navigation Menu Sidebar

### PR DESCRIPTION
**Pull request by gwwar**
Part of https://github.com/WordPress/gutenberg/issues/36667. 

This is a wide exploration PR which adds a new sidebar to the site and post editor to help with navigation menu management.

⚠️ **PLEASE NOTE:** A smaller, more targettted PR is now available in https://github.com/WordPress/gutenberg/pull/39290/ which focuses solely on the Navigation sidebar without addressing wider concerns about the "W" sidebar and global styles.

### After

The new navigation sidebar reuses the ListView component and allows users to easily reorder items even when the menu might be hidden (for example in a mobile menu).

https://user-images.githubusercontent.com/1270189/154540759-00d097a2-4231-4ab0-89ba-224ddc86b76f.mp4

### Before

Previously, clicking on the open list view button would open a modal.

| Open List View | Modal |
|-----|-----|
|  CleanShot 2022-02-10 at 13 44 02@2x | CleanShot 2022-02-10 at 13 43 00@2x |

### Testing Instructions

- Verify that in the Site Editor, opening list view now opens the new sidebar as expected
- We should see a loader when waiting for menu results
- Verify that in the Widget, Navigation and Post editor that the navigation sidebar is not available.

### Notes

Navigation blocks can store their data in [many ways](https://github.com/WordPress/gutenberg/blob/ca3c6dab4b0fbed1fb67d958a1ac1dd81609eb96/packages/block-library/src/navigation/use-navigation-entities.js):

1. With a reference to a wp_navigation post id
2. `__unstableLocation` such as a "primary" menu. The site editor incorrectly renders an empty placeholder in this case. Note that I won't address that in this PR.
3. With data set in innerBlocks.
4. A navigation block may also be populated from a "classic" menu

#### Wiring up the list view button on the Navigation block toolbar is a package puzzle.

- Each sidebar is namespaced to a particular editor instance, in case functionality needs to diverge. So for example the action we dispatch to open the sidebar in site editor is: 
```jsx
enableComplementaryArea(
			'core/edit-site',
			'edit-site/navigation-menu'
		); 
```
while in the post editor it is: 
```jsx
enableComplementaryArea(
			'core/edit-post',
			'edit-post/navigation-menu'
		);
```
- block-library should not know about a particular editor instance
- the toolbar in block-editor should not know about particular blocks

To solve this, I opted to add an unstable slot/fill following the pattern set by `__unstableBlockSettingsMenuFirstItem`. If folks have a better idea of how to untangle this, I'm happy to hear about suggestions.

### CODE REVIEW

**Comment by jasmussen** [comment]
Very cool. There are some excellent first steps to take in [this comment](https://github.com/WordPress/gutenberg/issues/36667#issuecomment-1030016317), and your PR and it feels like this one is blazing a trail towards that! 

On the side, I think we can explore some of the larger design-related changes with some fresh mockups, I'll get on that, but it's nothing that should block this functionality, so thanks for working on this!

> I'm open to early interaction feedback. For example @jasmussen if we added a select dropdown, what would that look like? Should we limit the listing to just the visible menus in the editor (in global blocks) or should we fetch all published menus?

In general I'm always a fan of starting as simple as possible and only add complexity on top of that when we learn that it's necessary. That could look like this:

* If there's only a single navigation stored, don't show the dropdown at all, just choose that.
* If there's multiple, default to showing the first one, and surface the dropdown letting you pick the others.

Buttons for creating and managing menus can always come later. But an additional benefit of starting with less is that it lets us figure out how prominent such controls need to be; is such menu management something you do all the time? Or do most sites indeed have a between zero and 2 menus in total?

Something to consider for the future is that case where a site has zero menus, and whether it might make sense to show all the pages of your site and only save it as a CPT if you make a change.

**Comment by gwwar** [comment]
> Something to consider for the future is that case where a site has zero menus, and whether it might make sense to show all the pages of your site and only save it as a CPT if you make a change.

Right, I think there's some interesting navigation menu building we can iterate on in future PRs. For the 0 navigation blocks case initially I might not show the sidebar, or perhaps just note that there are no navigation blocks (and maybe link to a shortcut for inserting one into the document).

> If there's only a single navigation stored, don't show the dropdown at all, just choose that.
> If there's multiple, default to showing the first one, and surface the dropdown letting you pick the others

Sounds good to me. I suppose we'll also see how useful the navigation menu names are. If we end up using it a lot I might need to rethink how we store data on the client to make the lookup a little less awkward.

One other thought is that folks will likely want to edit classic menus too, but that can also be left as a follow up. I imagine that we don't want to necessarily convert these navigation menus automatically.

**Comment by gwwar** [comment]
Not feeling great 🤧 so leaving some midday notes for myself:
- There's a straightforward action we can dispatch to enable a sidebar area (enableComplementaryArea), however these are namespaced to each editor. The navigation block is adding the list view button in the toolbar. Likely we'll remove this, and instead slot/fill from the editor, or find the appropriate hook to do so. One example is how we add a block settings menu item, but is using an unstable fill [__unstableBlockSettingsMenuFirstItem](https://github.com/WordPress/gutenberg/blob/bc3c87c0656a4867f88f6b689f4505742f3c79f4/packages/block-editor/src/components/block-settings-menu/block-settings-menu-first-item.js)
- For the dropdown we should probably display Navigation Menu names as select option labels. In some cases the blocks won't have a name (when it has innerBlocks). When they do have names, data access is cumbersome, since we need to fetch post items with getEntityRecords. It's worth looking into localizing data a bit, or creating another store.

**Comment by gwwar** [comment]
Heads up that I'm going to tidy up the summary

File: packages/block-editor/src/components/navigation-inspector/navigation-option.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
+/**
+ * WordPress dependencies
+ */
+import { useEntityProp } from '@wordpress/core-data';
```
Comment (tellthemachines) [inline packages/block-editor/src/components/navigation-inspector/navigation-option.js]: Is there any way we can avoid having `core-data` as a dependency of the block editor package? Block editor should be usable independently from WP.

Could we perhaps add the title as an attribute of the nav block, and then use `getBlock` to access it instead?

**Review by tellthemachines — COMMENTED** [review]
Thanks for trying this out! This is just a quick first pass from me, mainly testing; I haven't looked at the changeset in depth yet. 

A few thoughts/questions:

* Does it make sense to introduce this to the post editor, at least at this point? Because navs are being fetched from the editor content, and posts will most often not have them, it feels a bit unnecessary.
* Reading through #36667 it looks like the aim is for this sidebar to be a representation of the site structure, with each nav link actually taking us to a view of its page in the editor. This will only work if the nav consists solely of a curated list of links. Page list block is opaque, so if the nav is just a page list, we won't benefit from this view at all. And other blocks that might be added, e.g. Search or Spacer or even Social Icons will muddy that representation quite a bit. I don't have any ideas here except that we should probably decide early on whether we want this list to function as a nav-editing device, or a site structure representation, because we'll need quite different solutions for each of these aims.
* Wasn't this view meant to be on the left hand side of the editor? I noticed you also opened #38465, is that meant to complement this PR?
* The list view does provide a much neater interface to move nav items around ⭐ 
* We should probably look at a way to move items with the keyboard in list view (as a separate piece of work, of course 😅 )

**Comment by youknowriad** [comment]
Thanks for the PR! This looks like a great start.

I think my main remark here is close to what @tellthemachines shared above: basically this PR is making the added sidebar part of the "block editor" like an "additional block inspector" blocks can use to trigger sidebars while it seems to me like the main goal of the original issue is to add something that is more "site wide" and not specific to the current post/page/template being edited. In other words, this should be a sidebar of the site editor (and probably not have anything to do with complementary editor areas), it's more global that the thing being current edited in the site editor.

I think reusing the list view UI is indeed the right approach but it doesn't have to be the exact same thing used by the navigation block's modal, that one is contextual to the current navigation block being edited in the canvas.

**Comment by draganescu** [comment]
Nice work 👏🏻 

>  I imagine that we don't want to necessarily convert these navigation menus automatically.

I think we do. We don't want to manage classic menus via a block structure as this has resulted in many setbacks in last year's explorations for adding a block editor for wp menus.

**Comment by gwwar** [comment]
Thanks for the speedy feedback y'all, so to summarize what I'm hearing:

- [x] Keep this focused to the site editor for now. Perhaps pull this out from complementary areas
- [ ] It sounds like it'd be helpful to also add some exploratory commits with the new site structure of pinned icons. (I had been planning to leave that as a follow up for PR scope reduction). 
- [x] The new sidebar should be able to query the published list of menus. (Perhaps in the future we'd open this up to new menu creation/drafts as well?)
- [x] Think through how to populate a ListView or similar when the navigation block is not inserted in the edited template. 
- [ ] Data: I'd recommend this for a separate PR, but I'd agree that data lookup for navigation blocks are awkward, due to the multiple data sources and need to potentially fetch a core resource. We can push for more locality in our stores.
- [ ] It's also an interesting thought to show the children of page list in a ListView component, though we'd likely need to lock items down.
- [x] Keyboard support: turn on mover controls. As a follow up find an alternate way for moving items in list view if desired.

Questions on behavior:
- If I select a menu that is not inserted in one of the visible templates, should we render a preview of it in the main area?
- If we change the information architecture of the site editor, will this also affect the post, navigation and widget editor? What does it look like when a plugin adds pinned icons?
- The navigation block currently has an "open list view" button on the block toolbar. In the post editor, should we skip adding this button to the toolbar, or keep a modal view?

**Comment by gwwar** [comment]
So to move to fetching all navigation items, from the [package architecture page](https://developer.wordpress.org/block-editor/explanations/architecture/modularity/) it sounds like it might make sense to add some `wp_navigation` resource querying in the `@wordpresss/editor` package. Let me know if there's a better spot.

**Comment by gwwar** [comment]
Being able to load any menu gives us an interesting problem to solve. The ListView component currently assumes that any blocks it displays are available in the `block-editor` store. 

Imagine that we're editing the index template, which has a navigation block that points at a `wp_navigation` ref of `123`. If we select a menu with ref `456`, ListView will not be able to render that menu since the innerBlocks of that menu are not in the template.

There a few ways of approaching this, but I think these two might be more promising.

Option A: Provided that there's at least one navigation block in the template, selecting a menu ref, also updates the block attribute. I pushed a happy path example to the branch. There's still corner cases to work through.

https://user-images.githubusercontent.com/1270189/153677237-79a57ff3-a568-444f-8cab-189085276b04.mp4

Option B: Update ListView (or create a new component) to decouple dependence on specific data stores. If we pass a fully populated list of blocks, we should see a nice preview. @youknowriad had a preference for this one.
I can try drafting an alternate approach next week.

**Comment by jasmussen** [comment]
This is excellent:

dedicated nav

I've alluded to [some changes to where this lives](https://github.com/WordPress/gutenberg/pull/38600#issuecomment-1032350241) (such as moving the icons to the left) needing more exploration. But exploring in a panel such as this is an excellent way to make fast progress, so long as we keep in mind it could possibly move in the future.

Beyond those, some small thoughts inspired by the work in progress, but not necessary to solve as part of this PR:

- When moving items in the inspector, it'd be neat if the same animaton that happens with the mover control fire on the menu items. Possible? 
- We've discussed this in the past, that the list view easily gets wider than the allotted space, prompting horizontal scrollbars. Something to explore separately, and for all list views.

**Comment by youknowriad** [comment]
> So to move to fetching all navigation items, from the [package architecture page](https://developer.wordpress.org/block-editor/explanations/architecture/modularity/) it sounds like it might make sense to add some wp_navigation resource querying in the @wordpresss/editor package.

can't we just load `wp_navigation` records from `core/data` ?

> Option B: Update ListView (or create a new component) to decouple dependence on specific data stores. If we pass a fully populated list of blocks, we should see a nice preview. @youknowriad had a preference for this one.
I can try drafting an alternate approach next week.

I think we might not need much decoupling if we do something like this:


```
<BlockEditorProvider value={ [ navigatinBlockObject ] } onChange={ updateNavigationBlock }>
     <ListView />
</BlockEditorProvider>
```

Basically we're just telling the list view here to use its own block editor store with its own block list ...

In a sense the ListView is an "editor" on its own, it's just that it doesn't render `<BlockList>`  but renders another kind of "block list" (the tree)

File: packages/edit-site/src/components/sidebar/navigation-menu-sidebar/navigation-inspector.js [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+import { useState, useEffect, useCallback } from '@wordpress/element';
+import { SelectControl } from '@wordpress/components';
+import { store as coreStore } from '@wordpress/core-data';
+import {
+	__experimentalListView as ListView,
+	store as blockEditorStore,
+} from '@wordpress/block-editor';
+
+const EMPTY_BLOCKS = [];
+const NAVIGATION_MENUS_QUERY = [ { per_page: -1, status: 'publish' } ];
```
Comment (gwwar) [inline packages/edit-site/src/components/sidebar/navigation-menu-sidebar/navigation-inspector.js]: As a reminder note, we should probably think though paging and what it should look like. -1 will attempt to fetch all resources, which can be a problem when we have many menus. (page size is limited to 100).

**Comment by gwwar** [comment]
> can't we just load wp_navigation records from core/data

@youknowriad did you mean something like https://github.com/WordPress/gutenberg/pull/38600/commits/63e4dac9c3974020bdd02e9a66262a8b96540828 or using `getEntityRecords` without adding it to the default entities?

> Basically we're just telling the list view here to use its own block editor store with its own block list ...

Thanks, I'll explore that today!

> When moving items in the inspector, it'd be neat if the same animaton that happens with the mover control fire on the menu items. Possible?

@jasmussen The same animations should fire when using the movers. I suppose we could do something similar with a drag event. Note that the importance of this might change too, if we fully decouple the navigation pane from the template editor. 

Note that animations are disabled in the block list if we detect that folks prefer reduce motion, or if the editor has too many blocks that are loaded.

https://user-images.githubusercontent.com/1270189/153923977-74d3656d-948c-495f-98ff-ab3a2ad6be24.mp4

> We've discussed this in the past, that the list view easily gets wider than the allotted space, prompting horizontal scrollbars. Something to explore separately, and for all list views.

Yeah, this one is something to revisit. Once we get some good behavior outlined, I'm thinking that this panel can be feature flagged so we can continue UX explorations.

**Comment by gwwar** [comment]
Nice suggestion @youknowriad, I think that gets us 90% of the way using BlockEditorProvider! I just need to hunt down why dragging is broken. Most other functionality should behave as we expect

https://user-images.githubusercontent.com/1270189/153964265-d8590e4b-d941-41e2-bdae-b6f375391f56.mp4

**Comment by youknowriad** [comment]
> Nice suggestion @youknowriad, I think that gets us 90% of the way using BlockEditorProvider! I just need to hunt down why dragging is broken. Most other functionality should behave as we expect

Awesome

> @youknowriad did you mean something like https://github.com/WordPress/gutenberg/commit/63e4dac9c3974020bdd02e9a66262a8b96540828 or using getEntityRecords without adding it to the default entities?

I actually thought just using the default entities `getEntityRecords( 'postType', 'wp_navigation' )` if they are available.
I can see a point for having shortcuts.`getNavigationMenus` ... I know @getdave was also thinking about similar shortcuts for other entities (posts, pages, ...)

**Comment by gwwar** [comment]
I dug in a bit and dragging is not working at the moment, since we're missing `allowedBlocks` information that lets us know if blocks can be inserted as children or siblings. I'm looking into the way to populate this correctly, but feel free to drop a note if folks are more familiar. 

https://github.com/WordPress/gutenberg/blob/7cb61f9fab4d9bc6266e1e47aceed32f74cc91ff/packages/block-editor/src/components/list-view/use-list-view-drop-zone.js#L235-L243

**Comment by talldan** [comment]
> I dug in a bit and dragging is not working at the moment, since we're missing allowedBlocks information that lets us know if blocks can be inserted as children or siblings. I'm looking into the way to populate this correctly, but feel free to drop a note if folks are more familiar.

This is a similar problem to one of the challenges of navigation focus mode (https://github.com/WordPress/gutenberg/issues/37950). Focus mode would only render the inner blocks of the navigation block and not the navigation block itself, so it'd be missing the `allowedBlocks` settings. And also there would be no parent navigation block for the 'parent' setting on some of the child blocks.

There might be some extra challenges in List View. `allowedBlocks` is defined when the block `edit` is rendered via the `useInnerBlocksProps` hook, and blocks in list view don't actually render. It'll probably work ok if list view uses the same client ids as the main block editor. If we ever want to display a menu that isn't being rendered in the block editor there may be a problem.

Some brief past discussion about this kind of problem at the bottom of this comment - https://github.com/WordPress/gutenberg/pull/35418#issuecomment-944149096. Matias mentioned having a provider that could provide particular block list settings and the context of the parent block.

**Comment by gwwar** [comment]
Thanks for looking @talldan !

>  It'll probably work ok if list view uses the same client ids as the main block editor.

Yes, that works pretty well if we use the sidebar ListView to render the contents of the first navigation block or selected navigation block. The only downside to that is that we end up needing at least one navigation block in the site editor, and selecting another menu from the dropdown will update the ref of the navigation block. Folks can see this in https://github.com/WordPress/gutenberg/pull/38600/commits/58330ae78fc327840770961cc4eb36a23d6679b2

I also have a small workaround in https://github.com/WordPress/gutenberg/pull/38600/commits/a86243e1a47f9ab7841d487daacb4596e686ba47, where I populate `allowedBlocks` myself, but it's rather fragile.

> Matias mentioned having a provider that could provide particular block list settings and the context of the parent block

Interesting I can try and see if this is possible without too large of a change. One question I had was around different `allowedBlocks` per clientId (even if a block has the same type). I can maybe think of a few reasons why we might want this, but it'd be super appreciated if anyone happens to have a link to an issue or PR that I could look at for real use cases.

**Comment by getdave** [comment]
> I can see a point for having shortcuts.`getNavigationMenus` ... I know @getdave was also thinking about similar shortcuts for other entities (posts, pages, ...)

Yup. Still on my Todolist 😄

**Comment by gwwar** [comment]
Oh, hmm I did miss that we want to keep the pinned settings icon (and presumably plugin icons) on the right. I cherry-picked some commits in for a vertical icon strip, but will look into creating another slot/fill.

**Comment by gwwar** [comment]
I'll try to untangle the `allowedBlocks` issue tomorrow. I took a bit of time today to play around with a vertical placement of icons and start on some dark theme styling. Let me know if folks would like me to continue to exploring a vertical strip or would prefer icons being placed next to the W icon horizontally.

https://user-images.githubusercontent.com/1270189/154375650-a75a7ca9-1483-4ada-922e-a7dff7af966d.mp4

**Comment by gwwar** [comment]
Added some quick styling commits, so we can quickly see what this looks like with a darker sidebar theme. Some of these are temporary, for example we should ideally update components to add dark theme support and extend this via emotion for the components that implement that.

https://user-images.githubusercontent.com/1270189/154540759-00d097a2-4231-4ab0-89ba-224ddc86b76f.mp4

**Comment by gwwar** [comment]
I think there's enough here to get some general feedback on how this feels. A few questions on next steps:

- Are folks interested in continuing exploring the vertical dock/persistent strip? Otherwise I can start exploring other mocks in https://github.com/WordPress/gutenberg/issues/36667 such as moving the icons horizontally next to the site-logo/w-logo.
- Is there anything we'd like to change in terms of flow/behavior for the new navigation panel?

Once we're happy with the prototype, I can polish and pull out pieces for landing. For example, one smaller piece might be the new navigation panel that's hidden by a feature flag.

**Comment by jasmussen** [comment]
Took a look this morning. This is an excellent exploration that really helps explore the nooks and crannies of some of the concepts that have been discussed across a variety of tickets. Even just for uncovering those, this is much appreciated, nice work! There are a number of glitches and small issues which is to be expected, so I'll keep feedback high level for now.

Here's a quick GIF showing the navigation editing interface:

dedicated

* Conceptually it feels really good that _site wide changes_ live on the left, and _document specific changes_ live on the right. 
* Separating those two into the "dark" interface on the left, and the light interface on the right, helps further distinguish between "Site" and "Document".
* The change also makes the W menu much more valuable than it is today, and might even help inform future iterations of the larger interface.
* I can't completely figure out if reserving the slice of real estate on the left is a good or a bad idea. In practice it seems okay, but it still does take up valuable real estate. It does have the benefit of surfacing the tools with 1-click shortcuts, compared to having to first open the W menu. But another option might be to have the W menu be open by default as soon as you enter the site editor, and then choose whether you want to edit templates, navigation, or styles as your first choice. 
* The list view for global navigation editing continues to prove itself. I do think we need to reduce the dropdown for selecting the menu; it can still be easy to get to without being the top thing on the inspector. Not something to act on, just a note to self on designs. For example there might be a way to highlight/suggest an alternate menu if it happens to be loaded on the template you're looking at.

Diving into a GIF showing the style interface:
styles

* Layout glitches aside, styles also conceptually feel appropriate on the left for me. In terms of hierarchy, matching Site > Document > Metadata to the reading direction makes good sense.
* It does also surface, however, a shortcoming in terms of our components. If you drive into the "All blocks" section, we get the existing block inspector, only on the left insead of the right. In practice, this isn't as bad as I had expected, though! If we go in this direction we obviously need to make the components work in a dark canvas. But since editing "all blocks" is also a _site wide action_, the left might still feel like an appropriate place for this.

All in all, this direction to me feels like it's uncovering a slew of good opportunities. In terms of next steps, I'd love thoughts by @jameskoster on the high level conceptual stuff, but also @ciampo on the component side of things: if we went this direction, what would be the work load attached to making the inspector work on a dark background?

**Comment by jameskoster** [comment]
I agree that site-wide elements having their own space and colorisation feels very good.

> I can't completely figure out if reserving the slice of real estate on the left is a good or a bad idea.

This seems to be the key consideration here.

On the surface, single-click access to things like template management feels very sensible. But taking a step back, I think it could be worth exploring the impact that Browse mode (another 6.0 objective) has on these ideas before committing. 

It may be that a fluid pathway between browsing / editing creates a superior customisation experience for all of these things.

As an example it's probably more intuitive for most folks to navigate to a page on their site and edit from there, rather than expecting them to understand that the appearance of their homepage is governed by the "Index" template. If that is the case, and assuming we generally consider template management to be a power-user feature, then perhaps it's ok for it to be a couple of clicks away.

**Comment by jasmussen** [comment]
> As an example it's probably more intuitive for most folks to navigate to a page on their site and edit from there, rather than expecting them to understand that the appearance of their homepage is governed by the "Index" template. If that is the case, and assuming we generally consider template management to be a power-user feature, then perhaps it's ok for it to be a couple of clicks away.

Good thought.

I wonder if a starting point would be to start the site editor with the W menu open?

**Comment by jameskoster** [comment]
In theory, yes. With one-click access between browsing/managing, and editing. A rough mockup:


https://user-images.githubusercontent.com/846565/154963976-960e7ece-2777-45f9-90cd-990bd4ab93ab.mp4

**Comment by jasmussen** [comment]
"Rough mockup".

(I dig it.)

**Comment by ciampo** [comment]
Hey @jasmussen , thank you for the ping!

I just wanted to mention an experimental component that was recently added to the `@wordpress/components` package — [`Navigator`](https://wordpress.github.io/gutenberg/?path=/story/components-experimental-navigator--default). This component offers a way to declare "screens" and easily navigate between them (it is currently powering the Global Styles sidebar), without necessarily introducing any opinionated UI. It'd be interesting to see if it could be a good fit for this situation or, in case it isn't, understand the reason and see if we can iterate on it.

> if we went this direction, what would be the work load attached to making the inspector work on a dark background?

Currently components from `@wordpress/components` don't support theming out of the box, so in order to implement such a scenario, the consumer of the components will have to pass ad-hoc styles to override colors etc.

But theming is definitely an area that we intend to explore in the future, and the work on the Navigation Menu Sidebar could be very useful in exploring how that could look like.

**Comment by jasmussen** [comment]
> Currently components from @wordpress/components don't support theming out of the box, so in order to implement such a scenario, the consumer of the components will have to pass ad-hoc styles to override colors etc.

> But theming is definitely an area that we intend to explore in the future, and the work on the Navigation Menu Sidebar could be very useful in exploring how that could look like.

Thanks a bunch for chiming in. The missing theming is definitely a bit of a challenge for this PR, and even just the general concept of having inspector-like controls on a dark background, which is otherwise conceptually interesting. You can see [the GIF here](https://github.com/WordPress/gutenberg/pull/38600#issuecomment-1046619058) when diving into "All blocks", the controls start working less well there.

If we wanted a near-term solution to this, what's your best idea for accomplishing this? Color overrides in CSS could probably work, but we'd ideally want to do it in a way that didn't incur a ton of technical debt.

**Comment by gwwar** [comment]
> If we wanted a near-term solution to this, what's your best idea for accomplishing this? Color overrides in CSS could probably work

Zooming out a bit for theming requirements in components, I feel like there are a few different layers to this:

| os settings | Admin Theme | Plugin theme |
|-----|-----|-----|
| macos-big-sur-system-prefs-general-dark | component theming | CleanShot 2022-02-22 at 10 03 23@2x |

### OS Preferences for Dark/Light modes

If a user sets an OS level preference, the UI should ideally respond to this. If someone picks a dark mode, what does the current design look like for block editing areas?

### Admin theme scheme

Beyond that users may also pick an admin theme. `/wp-admin/profile.php`. We do use this a bit in the block editor, but this is primarily currently limited to the accent color. Do existing themes have enough colors in the pallet to differentiate between the global/site settings and the document?

### Plugin Theming

Finally, if `@wordpress/components` is used as a building block for building UI for plugins, it's highly likely that plugin developers will want to theme their UI with their product pallet. For example the yoast plugin may want to theme some components with `rgb(164, 40, 106)` background and `#fec228` as a primary

**Comment by jameskoster** [comment]
At the risk of sounding a bit boring, I think the Apple approach works very well. IE choose either light, dark, or auto mode, and then select an accent color to personalise.

Plugin theming feels like a whole other discussion. There's a question around the benefits of interfering with customer preferences in this way. Should branding be introduced another way?

**Comment by draganescu** [comment]
I wan to just post some 👍🏻 s for the cool separation of global dark left and local white right, feels like something that is quickly learned without training. 

In terms of navigation we should be able to create navigation in that area as well and I feel looking at the demos that we kinda need to introduce the concept of primary / secondary right there? Maybe, not sure.

>  it's probably more intuitive for most folks to navigate to a page on their site and edit from there, rather than expecting them to understand that the appearance of their homepage is governed by the "Index" template. 

@jameskoster this being totally true, it's also a thing that for a page to be homepage the index template is assigned, so the thing becomes should we have some default site structure where there is a menu that by default contains a "Home" link?

**Comment by jameskoster** [comment]
> should we have some default site structure where there is a menu that by default contains a "Home" link?

That's a good question. I'd love to hear more thoughts on this, but I would be inclined to exclude the home link by default because the Site Title / Logo generally provide this functionality.

The question then becomes: how do we expose pages that are not part of the default menu in the site structure? That's probably something to explore separately, but perhaps there's an "unlinked" section below the menu? That could contain things like the home page, and perhaps even post categories, tags, and other archives. Folks would be able to simply drag items from the "unlinked" section to their menu if they want to add them.

**Comment by ciampo** [comment]
> @jasmussen:
> The missing theming is definitely a bit of a challenge for this PR

We are definitely aware of that. Adding theming support (at least at a basic level) has already been discussed and we have an idea about how to approach it, although we haven't had the capacity to actually start experimenting on it yet (most of our time is spent helping folks with daily reviews, while slowly working on a few structural and fundamental package-wise improvements which will unblock future work).

We are still unsure about the specific requirements for the exposed theming APIs. In that regard, proceeding with style overrides on this project really is a helpful step because we can look at those overrides to understand what variables really need to be exposed, and we can keep that in mind so that the "theming" for this particular feature doesn't become too ambitious in the first place.

> @jasmussen:
> If we wanted a near-term solution to this, what's your best idea for accomplishing this? Color overrides in CSS could probably work, but we'd ideally want to do it in a way that didn't incur a ton of technical debt.

The best short-term solution would be to apply CSS overrides to these components where they are consumed (e.g. in the `@wordpress/block-editor` package), which we'd be removing once proper theming support is added natively to the components.

> @gwwar:
> Zooming out a bit for theming requirements in components, I feel like there are a few different layers to this 
>
> @jameskoster:
> At the risk of sounding a bit boring, I think the Apple approach works very well. IE choose either light, dark, or auto mode, and then select an accent color to personalise.
> Plugin theming feels like a whole other discussion. There's a question around the benefits of interfering with customer preferences in this way. Should branding be introduced another way?

The approach that was discussed on our end was to have the `@wordpress/components` package expose a few theming variables as CSS Custom properties.

It would be up to the consumer of `@wordpress/components` to establish the final value for each of these properties — for example, by default they could follow the "OS -> WP theme -> WP plugin" cascade mentioned above. But the Navigation Menu Sidebar could then override these settings so that every component rendered inside it would adopt a custom (dark) "theme".

**Comment by jasmussen** [comment]
> The question then becomes: how do we expose pages that are not part of the default menu in the site structure? That's probably something to explore separately, but perhaps there's an "unlinked" section below the menu? That could contain things like the home page, and perhaps even post categories, tags, and other archives. Folks would be able to simply drag items from the "unlinked" section to their menu if they want to add them.

That would be my instinct as well. There could be many pages, though, so we could need a button drawer for "more", so we at least avoid needing pagination.

> We are still unsure about the specific requirements for the exposed theming APIs. In that regard, proceeding with style overrides on this project really is a helpful step because we can look at those overrides to understand what variables really need to be exposed, and we can keep that in mind so that the "theming" for this particular feature doesn't become too ambitious in the first place.

👍 👍

I would echo Jay's assessment of Kerry's proposal, though: keep it simple, so components retain their baked-in accessibility and DNA. So working on light and dark backgrounds would be step 1. System level dark mode is probably not near term for the project, so that's mostly something to keep in mind, but even with that, there will be cases in light mode where we need components that show up on a dark background, so it can't just be a blanket switch.

For any theming properties beyond light and dark, I would imagine a single accent color and possibly border radius. Both should probably be component-wide, i.e. don't mix multiple accent colors and radiuses across the UI. Fonts and font sizes should probably be inherited by the context, mainly. Does that sound right to you?

**Comment by ciampo** [comment]
@jasmussen I agree with everything you said — we do want to keep things as simple as possible (especially during the first iterations), which includes exposing only the "theming variables" that strictly necessary. We'll definitely keep you, Jay and Kerry in mind for gather early feedback too!

**Comment by talldan** [comment]
I had a go at using inner blocks for Page List in https://github.com/WordPress/gutenberg/pull/39087, which would solve the issue of it not showing its contents in this sidebar.

It doesn't actually seem too bad. The new block locking features bring us a lot closer to this being a possibility.

**Comment by getdave** [comment]
This is great work.

I'd like to look at breaking this PR down to the essentials required to merge a Navigation Menu (only) sidebar. I'll be testing this PR and disentangling what is reusable.

**Comment by gwwar** [comment]
@getdave I'd be happy to pull out those parts for you.

**Comment by gwwar** [comment]
I rebased this branch to resolve conflicts. 

> I'd like to look at breaking this PR down to the essentials required to merge a Navigation Menu (only) sidebar.

@getdave I pulled this out for you in https://github.com/WordPress/gutenberg/pull/39290 feel free to commit directly if you head ideas.

**Comment by paaljoachim** [comment]
Testing using this method: https://make.wordpress.org/design/2021/03/03/testing-a-gutenberg-pull-request-pr/

It is quite a contrast.

Before:
Screenshot 2022-03-09 at 21 46 06

After:
Screenshot 2022-03-09 at 21 49 36

My initial reactions.
I have a vertical black bar in my layout. As in a sense it looks like it is part of my layout. I wonder why the Navigation and Styles are placed like they are. As earlier Styles icon was placed on the top right area. 
Clicking W I now see this:
Screenshot 2022-03-09 at 21 57 42
To me it becomes a bit confusing as the < Dashboard points toward the Nav and styles icon. The W menu area also becomes bigger.

What if the Nav and Styles were instead moved into the W menu area like so:
Moving Styles and Navigation into W menu

Here we are gradually recreating the left WordPress admin menu.

---

In addition to the adjustments I suggested above we could add the full Navigation into List view.

Currently we see this:
Screenshot 2022-03-09 at 22 07 16

What if we instead of the Page List actually saw when opening the Navigation in List view saw the full navigation and also moved the various links around. (Similar to how we can move things around in the current classic Navigation screen.)
List-view-Navigation-details

This way we are reusing an existing List view feature. Going into the details of a block. Being able to adjust the contents inside the block by using the List view.

----

Edit settings sidebar panel.
Adding a panel directly in the Navigation settings sidebar. Called for instance "Edit" to where one can manage the Navigation elements. Drag & Drop etc.
Edit-nav-in-sidebar

**Comment by getdave** [comment]
@paaljoachim Thank you for taking the time to review.

Unfortunately, _this_ PR is actually pretty broad right now and it trying to tackle a number of things. I can't see it landing in 6.0. 

Kerry ended up extracting a subset of this PR into https://github.com/WordPress/gutenberg/pull/39290 which adds a more simple Navigation-only sidebar.

I think we should pursue the more limited scope outlined in https://github.com/WordPress/gutenberg/pull/39290 else we're going to bite off more than we can chew.

It might be worth taking a look at that PR.

> add the full Navigation into List view.

This works already. You are using the Page List block which currently doesn't expose inner blocks within the list view. Instead try adding some standard links to the Navigation block and then opening it in list view. You'll see you have full control of the Nav structure within the list view.

That said I don't currently believe list view can handle all the requirements necessary to fully manage a Navigation (think managing styles, block settings, editing text .etc). It's going to end up replicating the functionality which we already have within the editor canvas via blocks. 

My preference would be for list view to remain "structure only" and for more dedicated approaches to handle the more complex parts of editing a Navigation.

**Comment by gwwar** [comment]
Goals in https://github.com/WordPress/gutenberg/issues/36667 are a bit broad. 

To clarify a bit further the PR here is a technical prototype that quickly shows what it looks like UX might feel like with a dedicated vertical rail on the left for global panels, and how we might make it easier to manage navigation structure.

I think folks are leaning toward other UX options, so it's pretty unlikely that this PR will land. 

I think it's still useful for folks to play with the branch, as it does quickly highlight some technical issues, especially if folks end up wanting to add additional functionality to ListView to make it more of a mini-editor. There's quite a number of side-effects that will need to be untangled from block list if that's so.

> What if we instead of the Page List actually saw when opening the Navigation in List view saw the full navigation and also moved the various links around.

@paaljoachim, @talldan explores this in https://github.com/WordPress/gutenberg/pull/39087

**Comment by paaljoachim** [comment]
A few things.
I was confused with seeing the Page list (as it become a kind of grouping of nav elements) and had to modify, Save the Navigation to have it open up to show the individual links. I found it weird to see the Page list as part of the navigation. It felt out of place being in the Navigation elements list.

From my limited exploration it seems that List view is working fairly well with drag and drop of individual links. It should be kept simple and deeper methods are needed elsewhere.

> I think we should pursue the more limited scope outlined in https://github.com/WordPress/gutenberg/pull/39290 else we're going to bite off more than we can chew.

I went ahead and left feedback there. Bottom line that I find would be helpful would be to in some way connect the "Navigation Editor" with the Nav block sidebar panels. Because a deeper editing of the Navigation should be associated with the Navigation settings.

**Comment by scruffian** [comment]
Created a rebase here: https://github.com/WordPress/gutenberg/pull/39885

**Pull request closed without merging**